File: src/test/jython/bitdepthConversion.py
# CLIJ example ImageJ Ops jython: bitdepthConversion.py
#
# This script shows how change the bit-depth of images in the GPU.
#
# Author: Deborah Schmidt, frauzufall@mpi-cbg.de
# May 2019
# ---------------------------------------------

#@ IOService io
#@ UIService ui
#@ OpService ops

from net.haesleinhuepf.clij import CLIJ

image64 = ops.create().img([20, 20])
image32 = ops.convert().float32(image64)
# TODO how to do that in a script?
#LoopBuilder.setImages(image32).forEachPixel(pixel -> pixel.set(pixel.getIndex()))
# init GPU
clij = CLIJ.getInstance()
# push image to GPU
inputGPU = clij.push(image32)

intensityScalingFactor = 255
temp = ops.run("multiplyImageAndScalarCLIJ", inputGPU, intensityScalingFactor)
image16 = ops.run("convertUInt16CLIJ", temp)
# show result
result = clij.pull(image16)
ui.show(result)
clij.close()

File: src/test/java/net/haesleinhuepf/clij/ops/examples/BitDepthConversion.java
package net.haesleinhuepf.clij.ops.examples;

import net.haesleinhuepf.clij.CLIJ;
import net.haesleinhuepf.clij.clearcl.ClearCLBuffer;
import net.haesleinhuepf.clij.coremem.enums.NativeTypeEnum;
import net.haesleinhuepf.clij.ops.reviewed.convertUInt16CLIJ.ConvertUInt16CLIJ;
import net.haesleinhuepf.clij.ops.reviewed.dilateBoxCLIJ.DilateBoxCLIJ;
import net.haesleinhuepf.clij.ops.reviewed.erodeBoxCLIJ.ErodeBoxCLIJ;
import net.haesleinhuepf.clij.ops.reviewed.multiplyImageAndScalarCLIJ.MultiplyImageAndScalarCLIJ;
import net.haesleinhuepf.clij.ops.reviewed.thresholdCLIJ.ThresholdCLIJ;
import net.imagej.ImageJ;
import net.imglib2.Dimensions;
import net.imglib2.FinalDimensions;
import net.imglib2.RandomAccessibleInterval;
import net.imglib2.img.Img;
import net.imglib2.loops.LoopBuilder;
import net.imglib2.type.numeric.real.DoubleType;
import net.imglib2.type.numeric.real.FloatType;

import java.io.IOException;

public class BitDepthConversion {

	public void run() {
		ImageJ ij = new ImageJ();
		//note that we cannot use DoubleType Images since they cannot be converted to ClearCLBuffer
		Img<DoubleType> image64 = ij.op().create().img(new long[]{20, 20});
		Img<FloatType> image32 = ij.op().convert().float32(image64);
		//TODO is this a ramp?
		LoopBuilder.setImages(image32).forEachPixel(pixel -> pixel.set(pixel.getIndex()));
		// init GPU
		CLIJ clij = CLIJ.getInstance();
		// push image to GPU
		ClearCLBuffer inputGPU = clij.push(image32);

		float intensityScalingFactor = 255;
		Object temp = ij.op().run(MultiplyImageAndScalarCLIJ.class, inputGPU, intensityScalingFactor);
		Object image16 = ij.op().run(ConvertUInt16CLIJ.class, temp);
		// show result
		RandomAccessibleInterval result = clij.convert(image16, RandomAccessibleInterval.class);
		ij.ui().show(result);
		clij.close();

	}

	public static void main(String ... args) throws IOException {
		BitDepthConversion task = new BitDepthConversion();
		task.run();
	}
}
